[SPARK-12941][SQL][MASTER] Spark-SQL JDBC Oracle dialect fails to map string datatypes to Oracle VARCHAR datatype

thomastechs opened the pull request:
## What changes were proposed in this pull request?

This Pull request is used for the fix SPARK-12941, creating a data type mapping to Oracle for the corresponding data type"Stringtype" from dataframe. This PR is for the master branch fix, where as another PR is already tested with the branch 1.4
## How was the this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
This patch was tested using the Oracle docker .Created a new integration suite for the same.The oracle.jdbc jar was to be downloaded from the maven repository.Since there was no jdbc jar available in the maven repository, the jar was downloaded from oracle site manually and installed in the local; thus tested. So, for SparkQA test case run, the ojdbc jar might be manually placed in the local maven repository(com/oracle/ojdbc6/11.2.0.2.0) while Spark QA test run.

thomastechs committed:
Adding new test suite for [SPARK-12941][SQL]

Added this new integration suite for creating the docker based test suite for the bug fix in SPARK-12941

thomastechs committed:
Update pom.xml for [SPARK-12941] docker test suite addition

Updated this pom.xml with the ojdbc jar related dependency to test the OracleIntegrationSuite related to SPARK-12941 bug fix

thomastechs committed:
Update DockerJDBCIntegrationSuite.scala for [SPARK-12941] docker test suite addition

Updated DockerJDBCIntegrationSuite.scala with the time out interval, as it is required to be atleast 200 sec for Oracle instance to be up

thomastechs committed:
Update OracleDialect.scala for [SPARK-12941] docker test suite addition

Updated OracleDialect.scala for [SPARK-12941]  for the fix of adding a data type mapping to Oracle

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] docker test suite addition

Updated  OracleIntegrationSuite.scala for [SPARK-12941]  with Oracle Docker test cases

yhuai commented:
ok to test

yhuai commented:
@thomastechs Thank you for the PR. Since the jar is not in maven repo and you have tested it locally, how about we ignore the test? Let's keep the test but just ignore it. Let's also comment how to run the test. What do you think?

AmplabJenkins commented:
Can one of the admins verify this patch?

yhuai commented:
ok to test

SparkQA commented:
**[Test build #51684 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51684/consoleFull)** for PR 11306 at commit [`bde06fa`](https://github.com/apache/spark/commit/bde06fa2189dfe9f9484d86a942eb6cfa50ffad8).

SparkQA commented:
**[Test build #51684 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51684/consoleFull)** for PR 11306 at commit [`bde06fa`](https://github.com/apache/spark/commit/bde06fa2189dfe9f9484d86a942eb6cfa50ffad8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51684/
Test FAILed.

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] Style check fixes and ignore test case

Updated the file with style check fixes and ignore test cases, as the ojdbc jar is not present in the maven.

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] Style check fixes

Style check fixes

SparkQA commented:
**[Test build #51717 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51717/consoleFull)** for PR 11306 at commit [`ca7ef4e`](https://github.com/apache/spark/commit/ca7ef4e3de03f70c7f37ede0075e31daf17ff4cc).

SparkQA commented:
**[Test build #51717 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51717/consoleFull)** for PR 11306 at commit [`ca7ef4e`](https://github.com/apache/spark/commit/ca7ef4e3de03f70c7f37ede0075e31daf17ff4cc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51717/
Test FAILed.

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] Style check fixes

Updated the order of imports

thomastechs commented:
@yhuai  Thanks for the comments. I have put the ignore instead of test. The steps for the testing would be as follows.
1. Pull oracle 11g image - docker pull wnameless/oracle-xe-11g
2. Start docker - sudo service docker start
3. Download oracle 11g driver jar and put it in maven local repo. (com/oracle/ojdbc6/11.2.0.2.0/ojdbc6-11.2.0.2.0.jar)
4. Run spark test - ./build/sbt "test-only org.apache.spark.sql.jdbc.OracleIntegrationSuite"

SparkQA commented:
**[Test build #51721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51721/consoleFull)** for PR 11306 at commit [`28a9e9f`](https://github.com/apache/spark/commit/28a9e9fc7793f0ae7dfc0fc0fae17d6ef8acd94e).

SparkQA commented:
**[Test build #51721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51721/consoleFull)** for PR 11306 at commit [`28a9e9f`](https://github.com/apache/spark/commit/28a9e9fc7793f0ae7dfc0fc0fae17d6ef8acd94e).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51721/
Test FAILed.

thomastechs commented:
@yhuai  The style checks have been passed. Currently the build fails because the maven dependency oracle jar is not available, as expected. Suppose, we comment the oracle jar dependency in the pom file, still the build might fail in the beforeAll step where it tries to create a connection to Oracle docker using the ojdbc jar(which is not present in SparkQA env). So what would you suggest to proceed?

yhuai commented on docker-integration-tests/pom.xml:
```diff
@@ -131,6 +131,12 @@
+      <artifactId>ojdbc6</artifactId>
+      <version>11.2.0.2.0</version>
+      <scope>test</scope>
+    </dependency>
```
Can you remove this change?

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala:
```diff
@@ -112,7 +112,7 @@ abstract class DockerJDBCIntegrationSuite
       docker.startContainer(containerId)
       jdbcUrl = db.getJdbcUrl(dockerIp, externalPort)
-      eventually(timeout(60.seconds), interval(1.seconds)) {
+      eventually(timeout(200.seconds), interval(200.seconds)) {
```
do we need to change this line?

thomastechs replied:
@yhuai  This line of timeout is required for the oracle related test.Because, the time taken by the oracle instance to bring up would not be enough with the old setting. We tested it in our local and  
confirmed this. If we do not change this time duration, Oracle would throw an exception, saying that the instance was tried to hit during the initialization stage.

yhuai replied:
Seems we need to have a smaller number of `interval`. Right now, the number of internal and timeout are the same.

thomastechs replied:
@yhuai    interval to be set as 200 sec, was a requirement to execute it fine in our local environment of oracle on docker.Less than that it causes exception related to early hit. Since the current test case is ignored, we may revert it back to 60,1; adding a comment here saying to increase this number, when testing with Oracle. Please let me know your thoughts.

yhuai replied:
Yea, sounds good. Let's revert this change and add a comment to `OracleIntegrationSuite`.

thomastechs replied:
Sure; A build is already in progress, I guess I can commit after that

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,57 @@
+import org.apache.spark.tags.DockerTest
+
+@DockerTest
+class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLContext {
```
Please add doc to explain how to run this suite.

yhuai replied:
Also, please add doc to explain why we have to ignore all tests of this suite.

thomastechs replied:
@yhuai  So, I shall add the comments in the OracleIntegrationSuite.scala itself on how to run ,version of ojdbc jar, and on why ignore. Hope this is what you meant by doc.

thomastechs committed:
Update pom.xml

Updating this pom.xml to comment the changes related oracle ojdbc jar dependency as the maven repository does not contain this.

SparkQA commented:
**[Test build #51741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51741/consoleFull)** for PR 11306 at commit [`87e3d7f`](https://github.com/apache/spark/commit/87e3d7f4de872852a3f16bf6a31761a8cfd1d46b).

SparkQA commented:
**[Test build #51741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51741/consoleFull)** for PR 11306 at commit [`87e3d7f`](https://github.com/apache/spark/commit/87e3d7f4de872852a3f16bf6a31761a8cfd1d46b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51741/
Test FAILed.

thomastechs committed:
Update OracleIntegrationSuite.scala

Updated the OracleIntegrationSuite.scala with the doc and override method

thomastechs committed:
Update OracleIntegrationSuite.scala

updated doc

SparkQA commented:
**[Test build #51762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51762/consoleFull)** for PR 11306 at commit [`250591d`](https://github.com/apache/spark/commit/250591dfb7c63f446cfb09702587642da2ec026d).

SparkQA commented:
**[Test build #51762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51762/consoleFull)** for PR 11306 at commit [`250591d`](https://github.com/apache/spark/commit/250591dfb7c63f446cfb09702587642da2ec026d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51762/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51760/
Test FAILed.

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] Style check fixes

Updated style check fixes

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51765/
Test FAILed.

thomastechs commented:
retest please. ok to test

yhuai commented:
test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51780/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #51782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51782/consoleFull)** for PR 11306 at commit [`d77d585`](https://github.com/apache/spark/commit/d77d5852b06c4003661757207868c693723210bd).

SparkQA commented:
**[Test build #51782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51782/consoleFull)** for PR 11306 at commit [`d77d585`](https://github.com/apache/spark/commit/d77d5852b06c4003661757207868c693723210bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51782/
Test PASSed.

yhuai commented on docker-integration-tests/pom.xml:
```diff
@@ -131,6 +131,14 @@
+        <version>11.2.0.2.0</version>
+        <scope>test</scope>
+      </dependency>
+    -->
```
Please also add comment at here to explain why we comment out this dependency.

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,87 @@
+
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types._
+import org.apache.spark.tags.DockerTest
```
Add a newline after this line.

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,87 @@
+ * "Stringtype" from dataframe. This PR is for the master branch fix, where as another PR
+ * is already tested with the branch 1.4
+ *
+ * This patch was tested using the Oracle docker.Created this integration suite for the same.
```
Whitespace before `Created`.

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,87 @@
+ * This integration suite is created as part of Pull request(11306) and is used for the fix
+ * SPARK-12941, creating a data type mapping to Oracle for the corresponding data type
+ * "Stringtype" from dataframe. This PR is for the master branch fix, where as another PR
+ * is already tested with the branch 1.4
```
Seems we do not need to have this paragraph.

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,87 @@
+ *    (com/oracle/ojdbc6/11.2.0.2.0/ojdbc6-11.2.0.2.0.jar)
+ * 4. Run spark test - ./build/sbt "test-only org.apache.spark.sql.jdbc.OracleIntegrationSuite"
+ *
+ */
```
We can mention that all tests in this suite are ignored and then delete the comment above the test case below.

thomastechs committed:
Update OracleIntegrationSuite.scala for [SPARK-12941] Updated Comments

Updated OracleIntegrationSuite.scala for [SPARK-12941] for the edits in the comments section

thomastechs committed:
Update pom.xml for [SPARK-12941] added comments

Added comments on why the oracle dependency was commented out

SparkQA commented:
**[Test build #51837 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51837/consoleFull)** for PR 11306 at commit [`b3de7f4`](https://github.com/apache/spark/commit/b3de7f493cc14e258cd699f395bd1241912beb4b).

SparkQA commented:
**[Test build #51837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51837/consoleFull)** for PR 11306 at commit [`b3de7f4`](https://github.com/apache/spark/commit/b3de7f493cc14e258cd699f395bd1241912beb4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51837/
Test PASSed.

thomastechs committed:
Update DockerJDBCIntegrationSuite.scala

Updated DockerJDBCIntegrationSuite.scala- Reverted the timeout and interval parameter to 60 and 1. This would be required to be increased for Oracle based testing.

thomastechs committed:
Updated OracleIntegrationSuite.scala

Updated the comments about the time interval change required in DockerJDBCIntegrationSuite.scala when testing with Oracle tests

SparkQA commented:
**[Test build #51849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51849/consoleFull)** for PR 11306 at commit [`3f0f3b8`](https://github.com/apache/spark/commit/3f0f3b8105b358f580b6225f9cf1592b44e17450).

SparkQA commented:
**[Test build #51849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51849/consoleFull)** for PR 11306 at commit [`3f0f3b8`](https://github.com/apache/spark/commit/3f0f3b8105b358f580b6225f9cf1592b44e17450).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51849/
Test FAILed.

jayadevanmurali commented:
@SparkQA Seems to be the build failed due kmeans error. I think this is not related to this PR. Could you please check

thomastechs commented:
@yhaui @SparkQA Looks like the build failed due to some other unit test cases, not related to this PR
Could you please confirm?

thomastechs commented:
@yhuai  Could you initiate a retest please?

yhuai commented:
test this please

SparkQA commented:
**[Test build #51918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51918/consoleFull)** for PR 11306 at commit [`3f0f3b8`](https://github.com/apache/spark/commit/3f0f3b8105b358f580b6225f9cf1592b44e17450).

SparkQA commented:
**[Test build #51918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51918/consoleFull)** for PR 11306 at commit [`3f0f3b8`](https://github.com/apache/spark/commit/3f0f3b8105b358f580b6225f9cf1592b44e17450).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51918/
Test PASSed.

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,80 @@
+    )
+    override val jdbcPort: Int = 1521
+    override def getJdbcUrl(ip: String, port: Int): String =
+      s"jdbc:oracle:thin:system/oracle@//$ip:$port/xe"
```
Sorry. Last question. If we leave these at here, we are still trying to establish the connection to the database through JDBC, right? Since we do not have jdbc driver, it will fail. Should we just comment out these code or just comment out the class?

yhuai commented on docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala:
```diff
@@ -0,0 +1,80 @@
+ * repository.
+ */
+@DockerTest
+class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLContext {
```
How about we use `Ignore` tag to ignore the entire suite? So, we should not even try to connect to oracle.

yhuai replied:
oh nvm. Before will not be triggered since no test will run.

jayadevanmurali replied:
@yhuai the beforeall method will not execute because there is no valid test inside this class(already put ignore on top of test).

yhuai commented:
Thanks. Merging to master.
